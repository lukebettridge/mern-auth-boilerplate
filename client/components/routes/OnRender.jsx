import { useEffect } from "react";
import PropTypes from "prop-types";

const OnRender = props => {
	useEffect(() => {
		if (props.method) props.method();
	}, []);
	return null;
};

OnRender.propTypes = {
	method: PropTypes.func
};

export default OnRender;
