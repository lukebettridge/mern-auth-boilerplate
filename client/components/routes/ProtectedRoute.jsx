import React from "react";
import PropType from "prop-types";
import { Route, Redirect } from "react-router-dom";
import Cookies from "js-cookie";

const ProtectedRoute = ({ component: Component, ...rest }) => {
	const isAuthenticated = () => {
		const authenticated = Cookies.get("authenticated");
		return authenticated && authenticated === "true";
	};

	return (
		<Route
			{...rest}
			render={props =>
				isAuthenticated() ? (
					<Component {...props} />
				) : (
					<Redirect
						to={{
							pathname: "/login",
							state: { from: props.location }
						}}
					/>
				)
			}
		/>
	);
};

ProtectedRoute.propTypes = {
	component: PropType.any,
	location: PropType.string
};

export default ProtectedRoute;
