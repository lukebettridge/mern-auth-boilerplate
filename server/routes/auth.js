const express = require("express");
const bcrypt = require("bcryptjs");
const jwt = require("jsonwebtoken");

const { mail, passwordHash, randomString } = require("../utils");

const validate = {
	register: require("../validation/register"),
	login: require("../validation/login"),
	resetPassword: require("../validation/reset-password")
};
const User = require("../src/models/User");
const RESET_KEY_STRING_LENGTH = 10;

const router = express.Router();

router.post("/register", (req, res) => {
	const { errors, isValid } = validate.register(req.body);
	if (!isValid) return res.status(400).json(errors);

	User.findOne({ email: req.body.email }).then(user => {
		if (user)
			return res.status(400).json({
				email:
					"The email address that you've entered is already associated with an account"
			});

		const newUser = new User({
			forename: req.body.forename,
			surname: req.body.surname,
			email: req.body.email,
			password: req.body.password
		});

		passwordHash(newUser.password, (err, hash) => {
			if (err) throw err;
			newUser.password = hash;
			newUser
				.save()
				.then(user => res.json(user))
				.catch(err => res.status(400).json(err));
		});
	});
});

router.post("/login", (req, res) => {
	const { errors, isValid } = validate.login(req.body);
	if (!isValid) return res.status(400).json(errors);

	const { email, password } = req.body;
	User.findOne({ email }).then(user => {
		if (!user)
			return res.status(404).json({
				email: "The email address that you've entered doesn't match any account"
			});

		bcrypt.compare(password, user.password).then(isMatch => {
			if (isMatch) {
				jwt.sign(
					{
						id: user.id,
						email: user.email
					},
					process.env.SECRET_OR_KEY,
					{
						expiresIn: "1d"
					},
					(_, token) => {
						res.cookie("jwt", token, {
							httpOnly: true,
							secure: process.env.NODE_ENV === "production"
						});
						res.json({ success: true });
					}
				);
			} else
				return res
					.status(400)
					.json({ password: "The password that you've entered is incorrect" });
		});
	});
});

router.get("/reset-password", (req, res) => {
	const { email } = req.query;
	if (!email)
		return res.status(400).json({
			email: "Email field is required"
		});

	User.findOne({ email }).then(user => {
		if (!user)
			return res.status(404).json({
				email: "The email address that you've entered doesn't match any account"
			});

		const date = new Date();
		date.setDate(date.getDate() + process.env.EXPIRE_PASSWORD_RESET_DAYS || 1);
		const resetKey = randomString(RESET_KEY_STRING_LENGTH, date);
		user
			.update({ resetKey })
			.then(() => {
				const url = `${process.env.BASE_URL}/auth/reset-password/${resetKey}`;
				mail({
					to: email,
					subject: `${process.env.APP_NAME} - Your password reset request`,
					text: `Hello ${user.forename},
					
					So, you've forgotten your password... no biggie, follow this link to set a new one: ${url}
					
					Thanks,
					The ${process.env.APP_NAME} Team`
				})
					.then(() => {
						res.json({ success: true });
					})
					.catch(err => res.status(400).json(err));
			})
			.catch(err => res.status(400).json(err));
	});
});

router.post("/reset-password", (req, res) => {
	const { errors, isValid } = validate.resetPassword(req.body);
	if (!isValid) return res.status(400).json(errors);

	const { newPassword, resetKey } = req.body;
	if (!resetKey)
		return res.status(400).json({
			resetKey: "No reset key was specified"
		});

	User.findOne({ resetKey }).then(user => {
		if (!user)
			return res.status(404).json({
				resetKey: "The reset key given doesn't match any account"
			});

		const milliseconds = Number(resetKey.substr(RESET_KEY_STRING_LENGTH));
		const date = new Date(milliseconds);
		if (new Date() > date)
			return res.status(400).json({
				resetKey: "The reset key given has expired"
			});

		passwordHash(newPassword, (err, hash) => {
			if (err) throw err;
			user
				.update({ password: hash, resetKey: undefined })
				.then(() => res.json({ success: true }))
				.catch(err => res.status(400).json(err));
		});
	});
});

module.exports = router;
