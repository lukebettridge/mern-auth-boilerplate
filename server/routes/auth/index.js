const router = require("express").Router();

router.post("/register", require("./register"));
router.post("/login", require("./login"));
router.post("/reset-password", require("./reset-password").post);
router.get("/reset-password", require("./reset-password").get);

module.exports = router;
