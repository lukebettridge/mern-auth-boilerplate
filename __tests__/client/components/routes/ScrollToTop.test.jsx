import React from "react";
import { mount } from "enzyme";
import toJSON from "enzyme-to-json";

import { ScrollToTop } from "components/routes/ScrollToTop";

global.window.scrollTo = jest.fn();
const history = { listen: jest.fn() };
const unlisten = jest.fn();

describe("ScrollToTop component", () => {
	beforeEach(() => {
		jest.resetAllMocks();
		history.listen.mockReturnValue(unlisten);
	});

	it("snapshot renders", () => {
		const subject = mount(<ScrollToTop history={history} />);
		expect(toJSON(subject)).toMatchSnapshot();
	});

	it("calls listen prop on mount", () => {
		const subject = mount(<ScrollToTop history={history} />);
		expect(history.listen).toHaveBeenCalledWith(expect.any(Function));
		expect(subject).toEqual({});
	});

	it("calls scrollTo on change of history", () => {
		mount(<ScrollToTop history={history} />);
		history.listen.mock.calls[0][0]();

		expect(global.window.scrollTo).toHaveBeenCalledWith(0, 0);
	});

	it("calls unlisten on unmount", () => {
		const subject = mount(<ScrollToTop history={history} />);
		subject.unmount();

		expect(unlisten).toHaveBeenCalled();
	});
});
