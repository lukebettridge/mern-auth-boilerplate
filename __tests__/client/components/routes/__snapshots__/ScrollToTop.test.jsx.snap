// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollToTop component snapshot renders 1`] = `
<ScrollToTop
  history={
    Object {
      "listen": [MockFunction] {
        "calls": Array [
          Array [
            [Function],
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": [MockFunction],
          },
        ],
      },
    }
  }
/>
`;
